### PR TITLE
Fixed for Issue #119/#121/#122/#123/#125

### PR DESCRIPTION
**Pull request by gebing**
Fixed for #119 Java.registerClass failed for interface with constants

### CODE REVIEW

**Comment by gebing** [comment]
Fixed for #121 Java.registerClass failed for interface with method of return byte[]

**Comment by oleavr** [comment]
Thanks! Could you also add a test for this fix? (To avoid accidentally breaking this in the future.)

**Comment by gebing** [comment]
OK. i will add some test

**Comment by gebing** [comment]
@oleavr 
I am trying to add some tests, but make failed by:

```
dx --dex --output=build/tests.dex build/tests.jar
make[1]: dx: No such file or directory
make[1]: *** [build/tests.dex] Error 1
make: *** [check] Error 2
```

what is `dx`?

**Comment by gebing** [comment]
Actually i am making the patch for #122, and already found a perfect solution. I just commit the code and still have a minor problem. My problem is in the implement method of new Java class, I do not have any solution to call the super class's method, such like below:

```
const superClass = Java.use("java.io.ByteArrayOutputStream");
const myOutputClass = Java.registerClass({
  name: 'com.example.MyOutputStream',
  superClass: "Ljava/io/ByteArrayOutputStream;",
  implements: [],
  methods: {
    write: [{
      returnType: 'void',
      argumentTypes: ['[B', 'int', 'int'],
      implementation: function (b, off, len) {
        console.log("MyOutputStream.write(" + JSON.stringify(b) + "," + off + "," + len + ")=" + Process.getCurrentThreadId());
        // Java.cast(this, superClass).write(b, off, len);
      }
    }, {
      returnType: 'void',
      argumentTypes: ['int'],
      implementation: function (b) {
        console.log('MyOutputStream.write(' + b + ")");
        // superClass.write.overload('int').call(this, b);
      }
    }]
  }
});
var myInstance = myOutputClass.$new();
Java.cast(myInstance, Java.use("java.io.OutputStream")).write(Java.use("java.lang.String").$new("123456").getBytes());
console.log("toString=" + myInstance.toString());
```

I have tried using `Java.cast` and `overload`, both methods will cause class's method is called recursively and make memory exhausted. Does any have any solution?

@oleavr

**Comment by gebing** [comment]
I add `$super` property to class to access super class's methods, and demo code is like below and all work fine:

```
global.myOutputClass = Java.registerClass({
  name: 'com.example.MyOutputStream',
  superClass: "java.io.ByteArrayOutputStream",
  implements: [],
  methods: {
    write: [{
      returnType: 'void',
      argumentTypes: ['[B', 'int', 'int'],
      implementation: function (b, off, len) {
        console.log("MyOutputStream.write(" + JSON.stringify(b) + "," + off + "," + len + ")");
        this.$super.write(b, off, len);
      }
    }, {
      returnType: 'void',
      argumentTypes: ['int'],
      implementation: function (b) {
        console.log('MyOutputStream.write(' + b + ")");
        this.$super.write(b);
      }
    }]
  }
});
var myInstance = myOutputClass.$new();
var bytes = Java.use("java.lang.String").$new("123456").getBytes();
Java.cast(myInstance, Java.use("java.io.OutputStream")).write(bytes);
Java.cast(myInstance, Java.use("java.io.OutputStream")).write(bytes);
console.log("toString=" + myInstance.toString());
myInstance = myOutputClass.$new();
myInstance.$super.write(bytes);
myInstance.$super.write(bytes);
console.log("toString=" + myInstance.toString());
myInstance = Java.use('com.example.MyOutputStream').$new();
Java.cast(myInstance, Java.use("java.io.OutputStream")).write(bytes);
Java.cast(myInstance, Java.use("java.io.OutputStream")).write(bytes);
console.log("toString=" + myInstance.toString());
```

@oleavr 
BTW: i still can not get JUnit to run, so i can not test my test code. can you help me to solve this?

**Comment by oleavr** [comment]
Did you try running it on an emulator, like an official 9.0 image from Google?

**Comment by gebing** [comment]
@oleavr 
Finally i finished all the bug fix and enhancement for Java.registerClass.

File: test/re/frida/MethodTest.java [hunk 1]
```diff
@@ -245,6 +246,15 @@ public void replacementCanBeDoneOnReflection() {
         }
     }
 
+    // Testcase for Issue #125
```
Comment (oleavr) [inline test/re/frida/MethodTest.java]:
```suggestion
    // Issue #125
```

File: test/re/frida/ClassCreationTest.java [hunk 2]
```diff
@@ -161,6 +168,147 @@ public void interfaceMethodCanHaveUnrelatedOverload() throws ClassNotFoundExcept
...
+                "});" +
+                "Java.use('re.frida.ClassCreationTest').myOutputClass.value = MyOutputStream.class;");
+        OutputStream myOutput = (OutputStream) myOutputClass.newInstance();
+        myOutput.write(new byte[] { '1', '2', '3', '4', '5' });
+        assertEquals("12345", myOutput.toString());
+        myOutput.write(new byte[] { '1', '2', '3', '4', '5' });
+        assertEquals("1234512345", myOutput.toString());
+    }
+
+    // Testcase for Issue #124
```
Comment (oleavr) [inline test/re/frida/ClassCreationTest.java]:
```suggestion
    // Issue #124
```

File: test/re/frida/ClassCreationTest.java [hunk 3]
```diff
@@ -161,6 +168,147 @@ public void interfaceMethodCanHaveUnrelatedOverload() throws ClassNotFoundExcept
...
+                "});" +
+                "Java.use('re.frida.ClassCreationTest').myOutputClass.value = MyOutputStream.class;");
+        OutputStream myOutput = (OutputStream) myOutputClass.newInstance();
+        myOutput.write(new byte[] { 1, 2, 3, 4, 5 });
+        assertEquals("1,2,3,4,5", myOutput.toString());
+        myOutput.write(new byte[] { 1, 2, 3, 4, 5 });
+        assertEquals("1,2,3,4,5,1,2,3,4,5", myOutput.toString());
+    }
+
+    // Testcase for Issue #122
```
Comment (oleavr) [inline test/re/frida/ClassCreationTest.java]:
```suggestion
    // Issue #122
```

File: test/re/frida/ClassCreationTest.java [hunk 4]
```diff
@@ -161,6 +168,147 @@ public void interfaceMethodCanHaveUnrelatedOverload() throws ClassNotFoundExcept
...
+                "    }," +
+                "  }" +
+                "});" +
+                "Java.use('re.frida.ClassCreationTest').primitiveArrayClass.value = SimplePrimitiveArray.class;");
+        PrimitiveArray primitiveArray = (PrimitiveArray) primitiveArrayClass.newInstance();
+        assertEquals(Arrays.equals(new byte[] { 1, 2, 3, 4, 5 }, primitiveArray.getByteArray()), true);
+        assertEquals("345", primitiveArray.setIntArray(new int[] { 1, 2, 3, 4, 5 }, 2));
+    }
+
+    // Testcase for Issue #122
```
Comment (oleavr) [inline test/re/frida/ClassCreationTest.java]:
```suggestion
    // Issue #122
```

File: test/re/frida/ClassCreationTest.java [hunk 5]
```diff
@@ -161,6 +168,147 @@ public void interfaceMethodCanHaveUnrelatedOverload() throws ClassNotFoundExcept
...
+                "  }" +
+                "});" +
+                "Java.use('re.frida.ClassCreationTest').hasFieldClass.value = SimpleHasField.class;");
+        HasField hasField = (HasField) hasFieldClass.newInstance();
+        assertEquals("hasField", hasField.getName());
+        assertEquals(100, hasField.getCalories(50));
+        assertEquals("Field", hasField.field);
+    }
+
+    // Testcase for Issue #121
```
Comment (oleavr) [inline test/re/frida/ClassCreationTest.java]:
```suggestion
    // Issue #121
```

File: test/re/frida/ClassCreationTest.java [hunk 6]
```diff
@@ -161,6 +168,147 @@ public void interfaceMethodCanHaveUnrelatedOverload() throws ClassNotFoundExcept
         assertEquals("checkServerTrusted B", script.getNextMessage());
     }
 
+    // Testcase for Issue #119
```
Comment (oleavr) [inline test/re/frida/ClassCreationTest.java]:
```suggestion
    // Issue #119
```

File: test/re/frida/ClassCreationTest.java [hunk 7]
```diff
@@ -5,6 +5,9 @@
 import org.junit.After;
 import org.junit.Test;
 
+import java.util.Arrays;
+import java.io.OutputStream;
+import java.io.ByteArrayOutputStream;
```
Comment (oleavr) [inline test/re/frida/ClassCreationTest.java]: These should be alphabetically sorted.

File: lib/class-factory.js [hunk 8]
```diff
@@ -1804,7 +1805,7 @@ function ClassFactory (vm) {
         const ifaceProto = Object.getPrototypeOf(iface);
         Object.getOwnPropertyNames(ifaceProto)
           .filter(name => {
-            return name[0] !== '$' && name !== 'constructor' && name !== 'class';
+            return name[0] !== '$' && name !== 'constructor' && name !== 'class' && iface[name].overloads;
```
Comment (oleavr) [inline lib/class-factory.js]:
```suggestion
            return name[0] !== '$' && name !== 'constructor' && name !== 'class' && iface[name].overloads !== undefined;
```

File: lib/class-factory.js [hunk 9]
```diff
@@ -1950,6 +1951,57 @@ function ClassFactory (vm) {
...
+                        return method.apply(superInstance, args);
+                      } finally {
+                        method[PENDING_CALLS].delete(tid);
+                      }
+                    }
+                  });
+                }
+                return new Proxy(prop, {
+                  apply: function (target, thisArg, args) {
+                    const method = prop.overloads.find(m => m.canInvokeWith(args));
```
Comment (oleavr) [inline lib/class-factory.js]: I believe `Array.prototype.find` is missing in the Duktape runtime, so this code presumably crashes there?

File: lib/class-factory.js [hunk 10]
```diff
@@ -1950,6 +1951,57 @@ function ClassFactory (vm) {
...
+                    if (!method) {
```
Comment (oleavr) [inline lib/class-factory.js]:
```suggestion
                    if (method === undefined) {
```

File: lib/class-factory.js (hunk 9 above)
Comment (gebing) [reply lib/class-factory.js]: Yes, you are right. But i wonder why test code does not crash, maybe it does not use Duktape runtime. I will change this code.

File: test/re/frida/ClassCreationTest.java (hunk 7 above)
Comment (gebing) [reply test/re/frida/ClassCreationTest.java]: OK, i will fix it.

File: lib/class-factory.js (hunk 9 above)
Comment (gebing) [reply lib/class-factory.js]: Fixed in lastest commit.

File: test/re/frida/ClassCreationTest.java (hunk 7 above)
Comment (gebing) [reply test/re/frida/ClassCreationTest.java]: Fixed in lastest commit.

**Comment by gebing** [comment]
@oleavr 
I have finished your change requests and commit the changes.

**Comment by oleavr** [comment]
Thanks!

**Pull request merged**